ee/secureenclavesigner/secureenclavesigner.go

```diff
@@ -0,0 +1,219 @@
+
+// Sign signs the digest using the secure enclave
+// If a public key is not set, it will create a new key
+func (ses *secureEnclaveSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	// create the key if we don't have it
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	if ses.pubKey == nil {
+		if err := ses.createKey(ctx); err != nil {
+			return nil, fmt.Errorf("creating key: %w", err)
+		}
+	}
+
+	pubKeyBytes, err := echelper.PublicEcdsaToB64Der(ses.pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("marshalling public key to der: %w", err)
+	}
+
+	signRequest := SignRequest{
+		Request: Request{
+			Challenge:    ses.challenge,
+			ServerPubKey: ses.serverPubKeyB64Der,
+		},
+		Digest:              digest,
+		SecureEnclavePubKey: pubKeyBytes,
+	}
+
+	signRequestMsgPack, err := msgpack.Marshal(signRequest)
+	if err != nil {
+		return nil, fmt.Errorf("marshalling sign request to msgpack: %w", err)
+	}
+
+	// get user name from uid
+	u, err := user.LookupId(ses.uid)
+	if err != nil {
+		return nil, fmt.Errorf("looking up user by uid: %w", err)
+	}
+
+	cmd := exec.CommandContext(ctx,
+		"/bin/launchctl",
+		"asuser",
+		ses.uid,
+		"sudo",
+		"--preserve-env",
+		"-u",
+		u.Username,
+		ses.pathToLauncherBinary,
+		"secure-enclave",
+		"sign",
+		base64.StdEncoding.EncodeToString(signRequestMsgPack),
+	)
+
+	// skip updates since we have full path of binary
+	cmd.Env = append(cmd.Environ(), fmt.Sprintf("%s=%s", "LAUNCHER_SKIP_UPDATES", "true"))
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("executing launcher binary to sign: %w: %s", err, string(out))
+	}
+
+	return []byte(lastLine(out)), nil
+}
+
+func (ses *secureEnclaveSigner) createKey(ctx context.Context) error {
+	// get user name from uid
+	u, err := user.LookupId(ses.uid)
+	if err != nil {
+		return fmt.Errorf("looking up user by uid: %w", err)
+	}
+
+	request := Request{
+		Challenge:    ses.challenge,
+		ServerPubKey: ses.serverPubKeyB64Der,
+	}
+
+	requestMsgPack, err := msgpack.Marshal(request)
+	if err != nil {
+		return fmt.Errorf("marshalling request to msgpack: %w", err)
+	}
+
+	cmd := exec.CommandContext(ctx,
+		"/bin/launchctl",
+		"asuser",
+		ses.uid,
+		"sudo",
+		"--preserve-env",
+		"-u",
+		u.Username,
+		ses.pathToLauncherBinary,
+		"secure-enclave",
+		"create-key",
+		base64.StdEncoding.EncodeToString(requestMsgPack),
+	)
+
+	// skip updates since we have full path of binary
+	cmd.Env = append(cmd.Environ(), fmt.Sprintf("%s=%s", "LAUNCHER_SKIP_UPDATES", "true"))
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("executing launcher binary to create key: %w: %s", err, string(out))
+	}
+
+	pubKey, err := echelper.PublicB64DerToEcdsaKey([]byte(lastLine(out)))
+	if err != nil {
+		return fmt.Errorf("marshalling public key to der: %w", err)
+	}
+
+	ses.pubKey = pubKey
+	return nil
+}
+
+func lastLine(out []byte) string {
```

Comment:
this feels a little iffy, but maybe feels better than adjusting main.go to skip logger

or maybe at the very beginning of main.go we see it's secure-enclave cmd and fire it off immediately.

Reply:
I think this is fine, but I'm also not opposed to killing the stderr logging in main.go in certain cases -- I think it's also unnecessarily verbose for e.g. `launcher version`, `launcher doctor`.